server/src/main/java/org/elasticsearch/cluster/metadata/DataStreamFailureStoreSettings.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.cluster.metadata;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.common.regex.Regex;
+import org.elasticsearch.common.settings.ClusterSettings;
+import org.elasticsearch.common.settings.Setting;
+
+import java.util.List;
+import java.util.function.Predicate;
+
+import static org.elasticsearch.core.Predicates.never;
+
+/**
+ * Holder for the data stream global settings relating to the data stream failure store. This defines, validates, and monitors the settings.
+ */
+public class DataStreamFailureStoreSettings {
+
+    private static final Logger logger = LogManager.getLogger(DataStreamFailureStoreSettings.class);
+
+    public static final Setting<List<String>> DATA_STREAM_FAILURE_STORED_ENABLED_SETTING = Setting.stringListSetting(
+        "data_streams.failure_store.enabled",
+        Setting.Property.Dynamic,
+        Setting.Property.NodeScope
+    );
+
+    private Predicate<String> failureStoreEnabledByName;
+
+    private DataStreamFailureStoreSettings() {
+        failureStoreEnabledByName = never();
+    }
+
+    /**
+     * Creates an instance and initialises the cluster settings listeners.
+     *
+     * @param clusterSettings The cluster settings to initialize the instance from and to watch for updates to
+     */
+    public static DataStreamFailureStoreSettings create(ClusterSettings clusterSettings) {
+        DataStreamFailureStoreSettings dataStreamFailureStoreSettings = new DataStreamFailureStoreSettings();
+        clusterSettings.initializeAndWatch(
+            DATA_STREAM_FAILURE_STORED_ENABLED_SETTING,
+            dataStreamFailureStoreSettings::setEnabledByNamePatterns
+        );
+        return dataStreamFailureStoreSettings;
+    }
+
+    /**
+     * Returns whether the settings indicate that the failure store should be enabled by the cluster settings for the given name.
+     *
+     * @param name The data stream name
+     * @param isInternal Whether this is an internal (system or dot-prefix) data stream: if true, this method will always return false
+     */
+    public boolean failureStoreEnabledForDataStreamName(String name, boolean isInternal) {
+        assert DataStream.isFailureStoreFeatureFlagEnabled() : "Testing whether failure store is enabled should be behind by feature flag";
+        return isInternal == false && failureStoreEnabledByName.test(name);
+    }
```

Comment:
I was a bit surprised to see the `isInternal` flag here. It feels to me like we are leaking information. I can walk you through how I was separating the concerns in my head and you can tell me if you agree or not.

The `DataStreamFailureStoreSettings` is only holding the failure store as default cluster setting, and since that's a list of regexes I think the predicate very well captures this. I think here, we shouldn't bother with the internal data streams.

The logic that this setting doesn't apply to internal data streams, I think, belongs with the resolution of the effective failure store enabled flag. There we pose the questions:
- "is the flag explicitly set?", if not,
- "is the data stream internal?", if not,
then we check with the settings. Does this make sense?

Reply:
Yeah, I totally see what you mean.

The reason I did it this way is that it would be a mistake to ever call `settings. failureStoreEnabledForDataStreamName(name)` without also doing the `isInternal` check. By putting the check in this method, we force the caller to think about it. Without, it would be possible for the caller to forget about that part of the requirements.

*However*, when I originally went through that thought process, I thought that we were going to be calling this method in multiple places. I've actually managed to common up the methods in `DataStream` so that it's only called from one place. So now my argument doesn't really hold water, as we have exactly one place we have to remember it either way 🙂.

So I think I'll take your advice and change it to the more intuitive thing.